ompi/mca/coll/spacc/coll_spacc_allreduce.c

```diff
@@ -0,0 +1,373 @@
+    int rank = ompi_comm_rank(comm);
+
+    OPAL_OUTPUT((ompi_coll_spacc_stream,
+                "coll:spacc:allreduce_intra_redscat_allgather: rank %d/%d",
+                rank, comm_size));
+
+    /* Find nearest power-of-two less than or equal to comm_size */
+    int nsteps = opal_hibit(comm_size, comm->c_cube_dim + 1);   /* ilog2(comm_size) */
+    int nprocs_pof2 = 1 << nsteps;                              /* flp2(comm_size) */
+
+    if (count < nprocs_pof2 || !ompi_op_is_commute(op)) {
+        OPAL_OUTPUT((ompi_coll_spacc_stream,
+                    "coll:spacc:allreduce_intra_redscat_allgather: rank %d/%d count %d switching to base allreduce",
+                    rank, comm_size, count));
+        return ompi_coll_base_allreduce_intra_basic_linear(sbuf, rbuf, count, dtype,
+                                                           op, comm, module);
+    }
+
+    int err = MPI_SUCCESS;
+    ompi_request_t *req[2];
+
+    ptrdiff_t lb, extent, dsize, gap = 0;
+    ompi_datatype_get_extent(dtype, &lb, &extent);
+    dsize = opal_datatype_span(&dtype->super, count, &gap);
+
+    /* Temporary buffer for receiving messages */
+    char *tmp_buf = NULL;
+    char *tmp_buf_raw = (char *)malloc(dsize);
+    if (NULL == tmp_buf_raw)
+        return OMPI_ERR_OUT_OF_RESOURCE;
+    tmp_buf = tmp_buf_raw - gap;
+
+    if (sbuf != MPI_IN_PLACE) {
+        /* Copy sbuf to rbuf */
+        err = ompi_datatype_copy_content_same_ddt(dtype, count, (char *)rbuf,
+                                                  (char *)sbuf);
+    }
+
+    /*
+     * Step 1. Reduce the number of processes to the nearest lower power of two
+     * p' = 2^{\floor{\log_2 p}} by removing r = p - p' processes.
+     * 1. In the first 2r processes (ranks 0 to 2r - 1), all the even ranks send
+     *    the second half of the input vector to their right neighbor (rank + 1)
+     *    and all the odd ranks send the first half of the input vector to their
+     *    left neighbor (rank - 1).
+     * 2. All 2r processes compute the reduction on their half.
+     * 3. The odd ranks then send the result to their left neighbors
+     *    (the even ranks).
+     *
+     * The even ranks (0 to 2r - 1) now contain the reduction with the input
+     * vector on their right neighbors (the odd ranks). The first r even
+     * processes and the p - 2r last processes are renumbered from
+     * 0 to 2^{\floor{\log_2 p}} - 1.
+     */
+
+    int vrank, step, wsize;
+    int nprocs_rem = comm_size - nprocs_pof2;
+
+    if (rank < 2 * nprocs_rem) {
+        int count_lhalf = count / 2;
+        int count_rhalf = count - count_lhalf;
+
+        if (rank % 2 != 0) {
+            /*
+             * Odd process -- exchange with rank - 1
+             * Send the left half of the input vector to the left neighbor,
+             * Recv the right half of the input vector from the left neighbor
+             */
+            err = MCA_PML_CALL(isend(rbuf, count_lhalf, dtype, rank - 1,
```

Comment:
```irecv() + send() + ompi_request_wait()``` is a bit faster that ```irecv() + isend() + ompi_request_wait_all(2, ...)```,
plus you do not have to worry about some error codes that are not supposed to be returned by collective.
last but not least, this is all automated in ```ompi_coll_base_sendrecv_actual()``` (or ```ompi_coll_base_sendrecv()``` if a MPI process might send/recv with itself